### PR TITLE
Add output plugin for Azure Log Analytics

### LINKED ISSUE #5317: Azure LogAnalytics Output Plugin
## Feature Request
Opening a feature request kicks off a discussion.

### Proposal:
Proposal for adding a new Output plugin for "Azure LogAnalytics" which will allow to send telegraph metrics over to Azure OMS portal.
https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api

### Use case: [Why is this important (helps with prioritizing requests)]
Not all users track metrics in AppInsights or AzureMonitor

### PR DESCRIPTION
**Pull request by yshahin**
Creating output plugin for Azure Log Analytics
Resolves #5317

### Required for all PRs:

- [✔️] Signed [CLA](https://influxdata.com/community/cla/).
- [✔️] Associated README.md updated.
- [✔️] Has appropriate unit tests.

### CODE REVIEW

**Comment by yshahin** [comment]
Who should close/merge this? Are we waiting for another reviewer?

**Comment by danielnelson** [comment]
I'll do a review during the 1.11 cycle

**Comment by gitlizko** [comment]
Any updates on this? This is a great addition and probably the need for this plugin will grow.

**Comment by gitlizko** [comment]
@yshahin @danielnelson 
Can you please merge this to next release. We have tested this with Linux and Windows machines and we're happy with the plugin. However the extra work to rebuild the installations for every release is frustrating. Thanks in advance.

**Comment by yshahin** [comment]
@danielnelson is there something I should do to have it merged? I thought I was waiting on you.
Thanks

File: README.md
Comment (danielnelson) [inline README.md:334]: Should we rename to `azure_monitor_logs` now?

Based on note on https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api

File: plugins/outputs/azure_loganalytics/doc.go
Comment (danielnelson) [inline plugins/outputs/azure_loganalytics/doc.go:3]: Remove this file, we aren't documenting packages with this style.

File: plugins/outputs/azure_loganalytics/azure_loganalytics.go
Comment (danielnelson) [inline plugins/outputs/azure_loganalytics/azure_loganalytics.go:142]: Wrap this line around 80 chars.  It seems like instead of RunCount we should be using len(reqBody), since the documentation says Content-Length are you sure this is correct?

File: plugins/outputs/azure_loganalytics/azure_loganalytics.go
Comment (danielnelson) [inline plugins/outputs/azure_loganalytics/azure_loganalytics.go:145]: Only return the error to avoid double logging.

File: plugins/outputs/azure_loganalytics/azure_loganalytics.go
Comment (danielnelson) [inline plugins/outputs/azure_loganalytics/azure_loganalytics.go:129]: Just an observations, this will be one request per metric name, which seems like it will be rather expensive.  I'm not sure what we can do about this, since we need to have different Log-Type to avoid the 500 field limitations.

File: plugins/outputs/azure_loganalytics/azure_loganalytics.go
Comment (danielnelson) [inline plugins/outputs/azure_loganalytics/azure_loganalytics.go:88]: It doesn't look like the context is used, you can just remove it.

File: plugins/outputs/azure_loganalytics/azure_loganalytics.go
Comment (danielnelson) [inline plugins/outputs/azure_loganalytics/azure_loganalytics.go:101]: Add `a.client.CloseIdleConnections()` here

**Comment by eehret** [comment]
I'm excited about this plugin, can't wait to see it get done!

**Comment by sjwang90** [comment]
@yshahin Looks like this plugin is waiting for a handful of updates based off of Daniel's comments. 

If you are also willing to submit this as an [external plugin](https://github.com/influxdata/telegraf/blob/master/docs/EXTERNAL_PLUGINS.md) that can be used with `execd` to run seamlessly with Telegraf that would be great. Telegraf users would be able to use the plugin sooner by having it as an external plugin since it wouldn't have to go through the typical review process. 

Feel free to respond with any questions you may have.

**Comment by yshahin** [comment]
@sjwang90 will take a look at the external plugin, would that mean I don't have to merge it here?

**Comment by sjwang90** [comment]
@yshahin Yes, this specific PR wouldn't be merged. We would simply ask you to submit a PR to [this readme](https://github.com/influxdata/telegraf/blob/master/EXTERNAL_PLUGINS.md#external-plugins) linking to the location of your plugin.

**Comment by yshahin** [comment]
@sjwang90 and how easy would it be to install is there a cli that would allow automatic download and install?

**Comment by sjwang90** [comment]
@yshahin We are planning on improving the build process for external plugins to run more adjacent with Telegraf itself - we'd love your feedback as we're building this out. 

Until then, we ask for the main readme to include clear [installation instructions](https://github.com/vipinvkmenon/awsalarms#installation)

**Comment by srebhan** [comment]
@yshahin are you still interested in getting this merged?

**Comment by yshahin** [comment]
I will apply these comment changes and move this to a plugin repo

**Comment by srebhan** [comment]
@yshahin we can also continue here. Please rebase your code and I'll give it a review. As you like...

File: plugins/outputs/azure_loganalytics/azure_loganalytics.go
Comment (yshahin) [reply plugins/outputs/azure_loganalytics/azure_loganalytics.go]: its required by createClient

File: plugins/outputs/azure_monitor_logs/README.md
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/README.md:3]:
```suggestion
This plugin sends metrics to Azure Log Analytics.
```

File: plugins/outputs/azure_monitor_logs/README.md
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/README.md:8]: I guess that's a copy'n paste leftover... ;-)

File: plugins/outputs/azure_monitor_logs/README.md
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/README.md:12]: Please add some documentation for the `shared_key` option.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:39]: Can we please make this a config option with this as default!?

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:55]: How about adding `TLSConfig` and maybe also the possibility of setting a proxy? Check the `http` input plugin. This should be the same for an output. Don't forget to add the config options.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:71]: Please document this limitation in the README and `sampleConfig`.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:78]: You might want to use the value of `namespacePrefixRegex` directly here instead of using a const.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:76]: In case zero is a valid timeout (which usually mean no-timeout), please rather specify this timeout in `init()` when creating the structure instance.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:83]: Please skip the `else` here as you return in the `if`
```suggestion
	if err != nil {
		return err
	} 
	if !match {
		return fmt.Errorf("namespace_prefix contains invalid characters")
	}
```

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:90]: Please remove the newline.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:126]: Please avoid newlines if they do not help visually structuring the code.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:148]: Newline.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:155]: Can we please use `strings.Join` here.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:167]: Can you please do this in `Write()` so we only deal with actually managing the sending in here!?

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:150]: Can you please do this in `Write()` so we only deal with actually managing the sending in here!?

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:45]: While being a matter of taste, I'm not a big fan of defining constants that are later used only once in the code and will _most likely_ never change. I leave it to you to keep those or move to the actual places where those are used (once).

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:181]: You probably should ready the body after checking for errors.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:202]: If you _really_ want this as an own function, please only return the client as this cannot produce an error.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:205]: Please remove newline.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:217]: Please remove newline.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:218]: Can we please do this in `Init()`. The `SharedKey` will not change right, so you can decode it once.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:240]: Please change to
```suggestion
		if tag.Key == "host" {
			m["Computer"] = tag.Value
			continue
		}
		m[underscoreToCaml(tag.Key)] = tag.Value
```

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs_test.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs_test.go:34]: This is equivalent to 
```suggestion
testutil.MustMetric(
	name,
	map[string]string{
		"host": "test-machine",
		"env":  "testing",
	},
	map[string]interface{}{"value": 42.0},
	time.Unix(0, 0),
)
```

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs_test.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs_test.go:40]: Maybe you should also test for the type/message of the error.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs_test.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs_test.go:49]: Same as above.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs_test.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs_test.go:82]: This will make an actual connection right? If so, please either setup a http-test-server (see e.g. http input plugin) or wrap this test so it is only executed in long tests, I prefer the first option.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs_test.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs_test.go:165]: Why not simply
```suggestion
			require.Error(t, err)
```

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs_test.go
Comment (srebhan) [inline plugins/outputs/azure_monitor_logs/azure_monitor_logs_test.go:240]: Please write one handler function and directly specify. Furthermore, you might want to test the received data outside of the http-test-server to simplify code.

**Review by srebhan — CHANGES_REQUESTED** [review]
Thanks @yshahin for submitting this code and sorry it took so long to review. Please see my comments in the code. My major requests are

1. Please make the service URL configurable. This will allow to react on changes in the service address and also eases testing.
2. Please consider adding TLS and Proxy configuration options.
3. Please use `testuil.MustMetric` for testing.
4. Please fix formatting! You can use `make fmt` to do it for you. Please note it will not fix the Readme or `sampleConfig`...

**Comment by srebhan** [comment]
@yshahin any news? Do you see a change to work on this PR?

**Comment by yshahin** [comment]
Sorry for the delay, I am working on it in my free time. Is there a deadline you would like to meet?

File: plugins/outputs/azure_loganalytics/azure_loganalytics.go
Comment (yshahin) [reply plugins/outputs/azure_loganalytics/azure_loganalytics.go]: We should not that and try to improve this with batching.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (yshahin) [reply plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:78]: I like that its in thee beginning of the file

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (yshahin) [reply plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:45]: I prefer to have those located at the top of the file.

File: plugins/outputs/azure_monitor_logs/azure_monitor_logs.go
Comment (srebhan) [reply plugins/outputs/azure_monitor_logs/azure_monitor_logs.go:83]: Still valid.

**Comment by yshahin** [comment]
I don't think I have the time to work on this. Also my team moved away from
this implementation.
If anyone is interested in continuing this please ping me and I can share
an info they need.

Other than that I will abandon this PR

Best Regards,
Youssef Shahin

**Comment by srebhan** [comment]
@yshahin thanks for letting us know. Sad to hear nevertheless...

**Comment by sjwang90** [comment]
Closing this PR due to inactivity. If someone is interested on taking this work back up, please feel free to re-open the PR and take over the code.

If you would still like this plugin, please submit it as an [external plugin](https://github.com/influxdata/telegraf/blob/master/docs/EXTERNAL_PLUGINS.md) that can be used with `execd` to run seamlessly with Telegraf.

**Pull request closed without merging**
